realm/src/main/java/io/realm/dynamic/DynamicRealmList.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.dynamic;
+
+import java.util.AbstractList;
+import java.util.Iterator;
+import java.util.ListIterator;
+
+import io.realm.Realm;
+import io.realm.internal.LinkView;
+
+/**
+ * RealmList exposed using a dynamic API. All objects in the list must have the same schema even though they are accessed
+ * dynamically.
+ */
+public class DynamicRealmList extends AbstractList<DynamicRealmObject> {
+
+    private final LinkView linkView;
+    private final Realm realm;
+
+    DynamicRealmList(LinkView linkView, Realm realm) {
+        this.linkView = linkView;
+        this.realm = realm;
+    }
+
+    @Override
+    public boolean add(DynamicRealmObject object) {
+        checkIsValidObject(object);
+        linkView.add(object.row.getIndex());
+        return true;
+    }
+
+    @Override
+    public void clear() {
+        linkView.clear();
+    }
+
+    @Override
+    public DynamicRealmObject get(int location) {
+        checkValidIndex(location);
+        return new DynamicRealmObject(realm, linkView.get(location));
+    }
+
+    @Override
+    public DynamicRealmObject remove(int location) {
+        DynamicRealmObject removedItem = get(location);
+        linkView.remove(location);
+        return removedItem;
+    }
+
+    @Override
+    public DynamicRealmObject set(int location, DynamicRealmObject object) {
+        checkIsValidObject(object);
+        checkValidIndex(location);
+        linkView.set(location, object.row.getIndex());
+        return object;
+    }
+
+    @Override
+    public int size() {
+        return ((Long)linkView.size()).intValue();
+    }
+
+    private void checkIsValidObject(DynamicRealmObject object) {
+        if (object == null) {
+            throw new IllegalArgumentException("DynamicRealmList does not accept null values");
+        }
+        if (!realm.getPath().equals(object.realm.getPath())) {
+            throw new IllegalArgumentException("Cannot add a object belonging already in another Realm");
+        }
+        if (!linkView.getTable().hasSameSchema(object.row.getTable())) {
+            String expectedClass = linkView.getTable().getName();
+            String objectClassName = object.row.getTable().getName();
+            throw new IllegalArgumentException("Object is of type " + objectClassName + ". Expected " + expectedClass);
+        }
+    }
+
+    private void checkValidIndex(int location) {
```

Comment:
everywhere else we use "check" for these checks - so keep that or change it everywhere else. "check" is better that "is" due to missing boolean return type. But "validateIndex" may be better than "checkValidIndex", but again requires that it's changed everywhere else. If that should be done it must be a separate PR.